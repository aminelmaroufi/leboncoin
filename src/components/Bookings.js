import React, { Component } from "react";
import { connect } from "react-redux";
import Map from "pigeon-maps";
import Marker from "pigeon-marker";
import Overlay from "pigeon-overlay";
import { getBookings } from "../actions/dash";

const mapStateToProps = (state, ownProps) => {
  return {
    bookings: state.dash.bookings,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onGetBookings: (params) => {
      dispatch(getBookings(params));
    },
  };
};

class Bookings extends Component {
  constructor(props) {
    super(props);
    this.state = {
      query: {
        price_min: 0,
        price_max: 3000,
        dep: "",
        furnished: false,
      },
      disabled: true,
    };
  }

  _handleChange = (att, value) => {
    let { query } = this.state;
    query[att] = value;

    this.setState({ query }, () => this.validateFeilds());
  };

  validateFeilds = () => {
    const { query } = this.state,
      price_min = parseInt(query.price_min),
      price_max = parseInt(query.price_max);
    let { disabled } = this.state;

    if (query.dep && price_min > 0 && price_max > price_min) {
      disabled = false;
    } else {
      disabled = true;
    }
    this.setState({ disabled });
  };

  search = () => {
    const { query } = this.state;
    this.props.onGetBookings(query);
  };

  render() {
    const { query, disabled } = this.state;
    const { bookings } = this.props;
    return (
      <div>
        {/*  filter */}
        <form style={{ paddingLeft: "20px" }}>
          <div className="" style={{ padding: "15px" }}>
            <div className="form-row align-items-center">
              <div className="col-2 ">
                <input
                  type="text"
                  className="form-control"
                  placeholder="Department"
                  value={query.dep}
                  onChange={(e) => this._handleChange("dep", e.target.value)}
                />
              </div>
              <div className="col-2 ">
                <input
                  type="text"
                  className="form-control"
                  placeholder="Price min"
                  value={query.price_min}
                  onChange={(e) =>
                    this._handleChange("price_min", e.target.value)
                  }
                />
              </div>
              <div className="col-2 ">
                <input
                  type="text"
                  className="form-control"
                  placeholder="Price max"
                  value={query.price_max}
                  onChange={(e) =>
                    this._handleChange("price_max", e.target.value)
                  }
                />
              </div>
              <div className="col-3" style={{ marginTop: "5px" }}>
                <div className="form-check">
                  <input
                    className="form-check-input"
                    type="checkbox"
                    id="autoSizingCheck2"
                    checked={query.furnished}
                    onChange={(e) =>
                      this._handleChange("furnished", !query.furnished)
                    }
                  />
                  <label
                    className="form-check-label"
                    htmlFor="autoSizingCheck2"
                    style={{ fontSize: "13px" }}
                  >
                    Furnished
                  </label>
                </div>
                <div className="">
                  <button
                    type="button"
                    className="btn btn-primary search-btn"
                    style={{
                      background: "white",
                      color: "#805ee4",
                      margin: "6px 0px",
                      float: "right",
                    }}
                    disabled={disabled}
                    onClick={() => this.search()}
                  >
                    Search for booking
                  </button>
                </div>
              </div>
            </div>
          </div>
        </form>
        <Map center={[48.864716, 2.349014]} zoom={12} width={600} height={400}>
          {bookings.length &&
            bookings.map((book) => {
              return (
                <Marker
                  key={book.id}
                  anchor={[book.location.lat, book.location.lng]}
                  payload={1}
                  onClick={({ event, anchor, payload }) => {}}
                />
              );
            })}

          <Overlay anchor={[48.864716, 2.349014]} offset={[120, 79]}>
            <img src="pigeon.jpg" width={240} height={158} alt="" />
          </Overlay>
        </Map>
      </div>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Bookings);
