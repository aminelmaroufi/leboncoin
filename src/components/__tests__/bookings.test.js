import React from "react";
import { MemoryRouter as Router, withRouter } from "react-router-dom";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { Bookings } from "../Bookings";

const search = jest.fn();

const max_price = "3,000€",
  dep = "Paris",
  price_min_value = "600€";

let renderer,
  form_title,
  dep_input,
  min_price_input,
  max_price_input,
  furnished_input,
  button_submit,
  details_title,
  map;

describe("BOOKINGS COMPONENTS", () => {
  beforeEach(() => {
    renderer = render(
      <Router>
        <Bookings onGetBookings={search} />
      </Router>
    );
    form_title = renderer.getByTestId("form-title");
    dep_input = renderer.getByTestId("dep-input");
    min_price_input = renderer.getByTestId("price-min");
    max_price_input = renderer.getByTestId("price-max");
    furnished_input = renderer.getByTestId("furnished-input");
    button_submit = renderer.getByTestId("submit-btn");
    details_title = renderer.getByTestId("submit-btn");
    map = renderer.getByTestId("map-container");
  });

  afterEach(cleanup);

  it("should render all form elements", () => {
    expect(form_title).toBeTruthy();
    expect(dep_input).toBeTruthy();
    expect(min_price_input).toBeTruthy();
    expect(max_price_input).toBeTruthy();
    expect(furnished_input).toBeTruthy();
    expect(button_submit).toBeTruthy();
    expect(map).toBeTruthy();
    expect(details_title).toBeTruthy();
    expect(renderer).toMatchSnapshot();
  });

  it("should update inputs value and call search event", () => {
    //check form input value is empty for each input
    expect(dep_input.value).toEqual("");
    expect(min_price_input.value).toEqual("0€");
    expect(max_price_input.value).toEqual(max_price);
    expect(furnished_input.value).toEqual("on");
    expect(button_submit.disabled).toBe(true);

    //fill form
    fireEvent.change(dep_input, { target: { value: dep } });
    fireEvent.change(min_price_input, { target: { value: price_min_value } });

    //check if input now are fiiled correctly
    expect(dep_input.value).toEqual(dep);
    expect(min_price_input.value).toEqual(price_min_value);
    expect(button_submit.disabled).toBe(false);

    fireEvent.click(button_submit);
    expect(search).toHaveBeenCalled();
    expect(search).toHaveBeenCalledTimes(1);
  });
});
