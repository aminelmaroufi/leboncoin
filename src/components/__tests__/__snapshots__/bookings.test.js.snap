// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BOOKINGS COMPONENTS should render all form elements 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          <form
            class="form_container"
            style="padding-left: 20px;"
          >
            <div
              class=""
              style="padding: 15px;"
            >
              <h5
                data-testid="form-title"
                style="margin-left: 25px;"
              >
                Filter bookings by :
              </h5>
              <div
                class="align-items-center"
              >
                <div
                  class="input-container"
                >
                  <input
                    class="form-control"
                    data-testid="dep-input"
                    placeholder="Department"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="input-container"
                >
                  <input
                    class="form-control"
                    data-testid="price-min"
                    placeholder="Price min"
                    type="text"
                    value="0"
                  />
                </div>
                <div
                  class="input-container"
                >
                  <input
                    class="form-control"
                    data-testid="price-max"
                    placeholder="Price max"
                    type="text"
                    value="3000"
                  />
                </div>
                <div
                  class="input-container"
                  style="margin-top: 5px;"
                >
                  <div
                    class="form-check"
                  >
                    <input
                      class="form-check-input"
                      data-testid="furnished-input"
                      id="autoSizingCheck2"
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      for="autoSizingCheck2"
                      style="font-size: 13px;"
                    >
                      Furnished
                    </label>
                  </div>
                </div>
                <div
                  class="btn_container"
                >
                  <button
                    class="btn btn-primary search-btn"
                    data-testid="submit-btn"
                    disabled=""
                    style="background: white; color: rgb(128, 94, 228); margin: 6px 0px; float: right;"
                    type="button"
                  >
                    Search for booking
                  </button>
                </div>
              </div>
            </div>
          </form>
          <div
            class="map_container"
            data-testid="map-container"
          >
            <div
              style="width: 600px; height: 400px; position: relative; display: inline-block; overflow: hidden; background: rgb(221, 221, 221);"
            >
              <div
                class=""
                style="width: 600px; height: 400px; position: absolute; top: 0px; left: 0px; overflow: hidden; will-change: transform; transform: scale(1, 1); transform-origin: top left;"
              >
                <div
                  style="position: absolute; width: 768px; height: 512px; will-change: transform; transform: translate(-141.99917795555666px, -55.99451764067635px);"
                >
                  <img
                    height="256"
                    loading="lazy"
                    src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2073/1408.png"
                    srcset=""
                    style="position: absolute; left: 0px; top: 0px; will-change: transform; transform-origin: top left; opacity: 1;"
                    width="256"
                  />
                  <img
                    height="256"
                    loading="lazy"
                    src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2073/1409.png"
                    srcset=""
                    style="position: absolute; left: 0px; top: 256px; will-change: transform; transform-origin: top left; opacity: 1;"
                    width="256"
                  />
                  <img
                    height="256"
                    loading="lazy"
                    src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2074/1408.png"
                    srcset=""
                    style="position: absolute; left: 256px; top: 0px; will-change: transform; transform-origin: top left; opacity: 1;"
                    width="256"
                  />
                  <img
                    height="256"
                    loading="lazy"
                    src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2074/1409.png"
                    srcset=""
                    style="position: absolute; left: 256px; top: 256px; will-change: transform; transform-origin: top left; opacity: 1;"
                    width="256"
                  />
                  <img
                    height="256"
                    loading="lazy"
                    src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2075/1408.png"
                    srcset=""
                    style="position: absolute; left: 512px; top: 0px; will-change: transform; transform-origin: top left; opacity: 1;"
                    width="256"
                  />
                  <img
                    height="256"
                    loading="lazy"
                    src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2075/1409.png"
                    srcset=""
                    style="position: absolute; left: 512px; top: 256px; will-change: transform; transform-origin: top left; opacity: 1;"
                    width="256"
                  />
                </div>
              </div>
              <div
                style="position: absolute; width: 600px; height: 400px; top: 0px; left: 0px;"
              >
                <div
                  class=""
                  style="position: absolute; transform: translate(180px, 121px);"
                >
                  <img
                    alt=""
                    height="158"
                    src="pigeon.jpg"
                    width="240"
                  />
                </div>
              </div>
              <div
                class="pigeon-attribution"
                style="position: absolute; bottom: 0px; right: 0px; font-size: 11px; padding: 2px 5px; background: rgba(255, 255, 255, 0.7); font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; color: rgb(51, 51, 51);"
              >
                <span>
                  <a
                    href="https://pigeon-maps.js.org/"
                    style="color: rgb(0, 120, 168); text-decoration: none;"
                  >
                    Pigeon
                  </a>
                   | 
                </span>
                <span>
                   Â© 
                  <a
                    href="https://www.openstreetmap.org/copyright"
                    style="color: rgb(0, 120, 168); text-decoration: none;"
                  >
                    OpenStreetMap
                  </a>
                   contributors
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="img_container"
        >
          <div
            style="margin-top: 20px;"
          >
            <br />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <form
          class="form_container"
          style="padding-left: 20px;"
        >
          <div
            class=""
            style="padding: 15px;"
          >
            <h5
              data-testid="form-title"
              style="margin-left: 25px;"
            >
              Filter bookings by :
            </h5>
            <div
              class="align-items-center"
            >
              <div
                class="input-container"
              >
                <input
                  class="form-control"
                  data-testid="dep-input"
                  placeholder="Department"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="input-container"
              >
                <input
                  class="form-control"
                  data-testid="price-min"
                  placeholder="Price min"
                  type="text"
                  value="0"
                />
              </div>
              <div
                class="input-container"
              >
                <input
                  class="form-control"
                  data-testid="price-max"
                  placeholder="Price max"
                  type="text"
                  value="3000"
                />
              </div>
              <div
                class="input-container"
                style="margin-top: 5px;"
              >
                <div
                  class="form-check"
                >
                  <input
                    class="form-check-input"
                    data-testid="furnished-input"
                    id="autoSizingCheck2"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="autoSizingCheck2"
                    style="font-size: 13px;"
                  >
                    Furnished
                  </label>
                </div>
              </div>
              <div
                class="btn_container"
              >
                <button
                  class="btn btn-primary search-btn"
                  data-testid="submit-btn"
                  disabled=""
                  style="background: white; color: rgb(128, 94, 228); margin: 6px 0px; float: right;"
                  type="button"
                >
                  Search for booking
                </button>
              </div>
            </div>
          </div>
        </form>
        <div
          class="map_container"
          data-testid="map-container"
        >
          <div
            style="width: 600px; height: 400px; position: relative; display: inline-block; overflow: hidden; background: rgb(221, 221, 221);"
          >
            <div
              class=""
              style="width: 600px; height: 400px; position: absolute; top: 0px; left: 0px; overflow: hidden; will-change: transform; transform: scale(1, 1); transform-origin: top left;"
            >
              <div
                style="position: absolute; width: 768px; height: 512px; will-change: transform; transform: translate(-141.99917795555666px, -55.99451764067635px);"
              >
                <img
                  height="256"
                  loading="lazy"
                  src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2073/1408.png"
                  srcset=""
                  style="position: absolute; left: 0px; top: 0px; will-change: transform; transform-origin: top left; opacity: 1;"
                  width="256"
                />
                <img
                  height="256"
                  loading="lazy"
                  src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2073/1409.png"
                  srcset=""
                  style="position: absolute; left: 0px; top: 256px; will-change: transform; transform-origin: top left; opacity: 1;"
                  width="256"
                />
                <img
                  height="256"
                  loading="lazy"
                  src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2074/1408.png"
                  srcset=""
                  style="position: absolute; left: 256px; top: 0px; will-change: transform; transform-origin: top left; opacity: 1;"
                  width="256"
                />
                <img
                  height="256"
                  loading="lazy"
                  src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2074/1409.png"
                  srcset=""
                  style="position: absolute; left: 256px; top: 256px; will-change: transform; transform-origin: top left; opacity: 1;"
                  width="256"
                />
                <img
                  height="256"
                  loading="lazy"
                  src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2075/1408.png"
                  srcset=""
                  style="position: absolute; left: 512px; top: 0px; will-change: transform; transform-origin: top left; opacity: 1;"
                  width="256"
                />
                <img
                  height="256"
                  loading="lazy"
                  src="https://stamen-tiles.a.ssl.fastly.net/toner/12/2075/1409.png"
                  srcset=""
                  style="position: absolute; left: 512px; top: 256px; will-change: transform; transform-origin: top left; opacity: 1;"
                  width="256"
                />
              </div>
            </div>
            <div
              style="position: absolute; width: 600px; height: 400px; top: 0px; left: 0px;"
            >
              <div
                class=""
                style="position: absolute; transform: translate(180px, 121px);"
              >
                <img
                  alt=""
                  height="158"
                  src="pigeon.jpg"
                  width="240"
                />
              </div>
            </div>
            <div
              class="pigeon-attribution"
              style="position: absolute; bottom: 0px; right: 0px; font-size: 11px; padding: 2px 5px; background: rgba(255, 255, 255, 0.7); font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; color: rgb(51, 51, 51);"
            >
              <span>
                <a
                  href="https://pigeon-maps.js.org/"
                  style="color: rgb(0, 120, 168); text-decoration: none;"
                >
                  Pigeon
                </a>
                 | 
              </span>
              <span>
                 Â© 
                <a
                  href="https://www.openstreetmap.org/copyright"
                  style="color: rgb(0, 120, 168); text-decoration: none;"
                >
                  OpenStreetMap
                </a>
                 contributors
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="img_container"
      >
        <div
          style="margin-top: 20px;"
        >
          <br />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
